Source/Urho3D/Scene/ScenePrefab.h

```diff
@@ -0,0 +1,155 @@
+class URHO3D_API AttributePrefab
+{
+public:
+    enum class IdentifierType : unsigned char
+    {
+        Id,
+        Name,
+        NameHash,
+        Unused,
+        // Don't add more types!
+    };
+
+    AttributePrefab() = default;
+    explicit AttributePrefab(AttributeId id);
+    explicit AttributePrefab(const ea::string& name);
+    explicit AttributePrefab(const char* name);
+    explicit AttributePrefab(StringHash nameHash);
+
+    void SetValue(const Variant& value) { value_ = value; }
+    void SetValue(Variant&& value) { value_ = ea::move(value); }
+
+    void SerializeInBlock(Archive& archive, bool compactSave = false);
+
+    IdentifierType GetIdentifierType() const;
+
+    AttributeId GetId() const { return id_; }
+    const ea::string& GetName() const { return name_; }
+    StringHash GetNameHash() const { return nameHash_; }
+    VariantType GetType() const { return value_.GetType(); }
+    const Variant& GetValue() const { return value_; }
+
+    bool operator==(const AttributePrefab& rhs) const;
+    bool operator!=(const AttributePrefab& rhs) const { return !(*this == rhs); }
+
+private:
+    AttributeId id_{};
+    ea::string name_;
+    StringHash nameHash_;
+
+    Variant value_{};
+};
+
+URHO3D_API void SerializeValue(Archive& archive, const char* name, AttributePrefab& value, bool compactSave = false);
+
+/// Serializable prefab.
+/// Contains a list of attributes.
+class URHO3D_API SerializablePrefab
+{
+public:
+    SerializablePrefab() = default;
+
+    void SetType(const ea::string& typeName);
+    void SetType(const char* typeName);
+    void SetType(StringHash typeNameHash);
+    void SetId(SerializableId id) { id_ = id; }
+
+    void Import(const Serializable* serializable, PrefabSaveFlags flags = {});
+    void Export(Serializable* serializable, PrefabLoadFlags flags = {}) const;
+
+    void SerializeInBlock(Archive& archive, PrefabArchiveFlags flags = {}, bool compactSave = false);
+
+    AttributeScopeHint GetEffectiveScopeHint(Context* context) const;
+
+    const ea::string& GetTypeName() const { return typeName_; }
+    StringHash GetTypeNameHash() const { return typeNameHash_; }
+    SerializableId GetId() const { return id_; }
+    const ea::vector<AttributePrefab>& GetAttributes() const { return attributes_; }
+    ea::vector<AttributePrefab>& GetMutableAttributes() { return attributes_; }
+
+    bool operator==(const SerializablePrefab& rhs) const;
+    bool operator!=(const SerializablePrefab& rhs) const { return !(*this == rhs); }
+
+private:
+    ea::string typeName_;
+    StringHash typeNameHash_;
+
+    SerializableId id_{};
+    bool temporary_{};
+
+    ea::vector<AttributePrefab> attributes_;
+};
+
+URHO3D_API void SerializeValue(Archive& archive, const char* name, SerializablePrefab& value,
+    PrefabArchiveFlags flags = {}, bool compactSave = false);
+
+/// Scene prefab.
+/// Contains node attributes, components and child nodes.
+class URHO3D_API ScenePrefab
```

Comment:
it's a bit strange that the class is named ScenePrefab although it usually contains data to recreate Node. "Scene" as it belongs to Scene? Maybe NodePrefab would reflect the purpose of the class better.

Reply:
It contains data to recreate Node hiearchy with components, which *is* the Scene. Or the part of the scene.
I'll think about rename tho.

Reply:
It looks to me that every node in hierarchy is represented by ScenePrefab, hence the question.